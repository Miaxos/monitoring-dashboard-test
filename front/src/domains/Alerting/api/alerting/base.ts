export type AlertingEventBase = {
  date: Date;
};
