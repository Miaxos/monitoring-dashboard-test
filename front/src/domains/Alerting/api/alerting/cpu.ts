import { createOpaqueAPI } from '@iadvize-oss/opaque-type';
import { pipe } from 'fp-ts/es6/pipeable';
import { AlertingEventBase } from './base';

enum EventType {
  CPU_ALERT = 'CPU_ALERT',
  CPU_RECOVER = 'CPU_RECOVER',
}

type $CPUEvent = { type: EventType } & AlertingEventBase;
const CPUEvent = createOpaqueAPI<EventType.CPU_ALERT, $CPUEvent>(
  EventType.CPU_ALERT,
);

export type CPUEvent = ReturnType<typeof CPUEvent.toOpaque>;

/**
 * Create a CPU Alert
 */
export function alert(creationDate: Date): CPUEvent {
  return CPUEvent.toOpaque({
    type: EventType.CPU_ALERT,
    date: creationDate,
  });
}

/**
 * Create a CPU Recover
 */
export function recover(creationDate: Date): CPUEvent {
  return CPUEvent.toOpaque({
    type: EventType.CPU_RECOVER,
    date: creationDate,
  });
}

/**
 * Return the date from a CPUEvent
 */
export function date(event: CPUEvent): Date {
  return pipe(event, CPUEvent.fromOpaque, elt => elt.date);
}

/**
 * Pattern matching on a CPUEvent
 */
export function fold<R>({
  onAlert,
  onRecover,
}: {
  onAlert: (event: CPUEvent) => R;
  onRecover: (event: CPUEvent) => R;
}) {
  return (event: CPUEvent) =>
    pipe(event, CPUEvent.fromOpaque, elt => {
      switch (elt.type) {
        case EventType.CPU_ALERT:
          return onAlert(event);
        case EventType.CPU_RECOVER:
          return onRecover(event);
      }
    });
}
