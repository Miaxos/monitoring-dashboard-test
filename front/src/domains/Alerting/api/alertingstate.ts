import * as ArrayFP from 'fp-ts/es6/Array';
import * as Option from 'fp-ts/es6/Option';
import { createOpaqueAPI } from '@iadvize-oss/opaque-type';
import { pipe } from 'fp-ts/es6/pipeable';
import * as CPUEvent from './alerting/cpu';

type $AlertingState = {
  alertingEvents: CPUEvent.CPUEvent[];
  startedTime: Date;
};

const AlertingState = createOpaqueAPI<'ALERTING_STATE', $AlertingState>(
  'ALERTING_STATE',
);
export type AlertingState = ReturnType<typeof AlertingState.toOpaque>;

/**
 * Create alerting state
 */
export function create(
  alertingEvents: CPUEvent.CPUEvent[] = [],
): AlertingState {
  return AlertingState.toOpaque({
    alertingEvents,
    startedTime: new Date(Date.now()),
  });
}

/**
 * Add an event to the alerting state
 */
export function add(event: CPUEvent.CPUEvent) {
  return (state: AlertingState) =>
    pipe(
      state,
      AlertingState.fromOpaque,
      elt => ({ ...elt, alertingEvents: [event, ...elt.alertingEvents] }),
      AlertingState.toOpaque,
    );
}

/**
 * Get the events from a AlertingState
 * Events aren't sorted by date, but the last in is the first.
 */
export function events(state: AlertingState): CPUEvent.CPUEvent[] {
  return pipe(state, AlertingState.fromOpaque, elt => elt.alertingEvents);
}

/**
 * Get the started time of the alerting
 */
export function startedSince(state: AlertingState): Date {
  return pipe(state, AlertingState.fromOpaque, elt => elt.startedTime);
}

/**
 * It represent the Alerting actual state, either Normal or with an Alert
 */
export enum AlertingStatus {
  NORMAL = 'NORMAL',
  ALERT = 'ALERT',
}

type StatusResponse = {
  status: AlertingStatus;
  date: Option.Option<Date>;
};

/**
 * Get the actual status of the Alerting State and the time when it started.
 */
export function status(state: AlertingState): StatusResponse {
  return pipe(
    state,
    events,
    ArrayFP.head,
    Option.fold(
      () => ({
        status: AlertingStatus.NORMAL,
        date: Option.none,
      }),
      event =>
        pipe(
          event,
          CPUEvent.fold<StatusResponse>({
            onAlert: alertEvent => ({
              status: AlertingStatus.ALERT,
              date: Option.some(CPUEvent.date(alertEvent)),
            }),
            onRecover: recoverEvent => ({
              status: AlertingStatus.NORMAL,
              date: Option.some(CPUEvent.date(recoverEvent)),
            }),
          }),
        ),
    ),
  );
}
