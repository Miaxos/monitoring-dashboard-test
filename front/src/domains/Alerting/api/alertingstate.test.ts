import * as Option from 'fp-ts/es6/Option';
import * as AlertingState from './alertingstate';
import * as CPUEvent from './alerting/cpu';

describe('domains/Alerting/api/alertingstate', () => {
  it('should return a good projection of the status when we have alerts', () => {
    const orderedDate = [
      '2020-03-01T22:24:32.500Z',
      '2020-03-02T22:24:32.500Z',
      '2020-03-03T22:24:32.500Z',
      '2020-03-04T22:24:32.500Z',
      '2020-03-05T22:24:32.500Z',
      '2020-03-06T22:24:32.500Z',
      '2020-03-07T22:24:32.500Z',
      '2020-03-08T22:24:32.500Z',
    ].map((elt, index) =>
      index % 2 === 0
        ? CPUEvent.alert(new Date(elt))
        : CPUEvent.recover(new Date(elt)),
    );

    const state = AlertingState.create(orderedDate);
    const status = AlertingState.status(state);

    expect(status).toMatchSnapshot();
  });

  it('should return a good projection of the status when we have no alerts', () => {
    const state = AlertingState.create();
    const status = AlertingState.status(state);

    expect(status).toStrictEqual({
      status: AlertingState.AlertingStatus.NORMAL,
      date: Option.none,
    });
  });
});
