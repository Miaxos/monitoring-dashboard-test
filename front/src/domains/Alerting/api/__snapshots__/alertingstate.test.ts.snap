// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`domains/Alerting/api/alertingstate should return a good projection of the status when we have alerts 1`] = `
Object {
  "date": Object {
    "_tag": "Some",
    "value": 2020-03-01T22:24:32.500Z,
  },
  "status": "ALERT",
}
`;
